enhancements/update/configurable-clusterversion-signature-stores.md

```diff
@@ -0,0 +1,197 @@
+
+Currently, there is no way for the CVO to fetch the signatures through custom URIs. This enhancement defines an API change in the CVO
+to accept a custom update-service URI which can be used to verify release signatures. 
+
+
+### User Stories
+
+* As a cluster administrator, I want the cluster's CVO to add new signature store URIs, so that I can have local release signature 
+  verification in air gapped environments.
+
+### Goals
+
+* To verify the release image signature in disconnected environments using OSUS signatures URI.
+* To replace existing signature stores with custom user-provided stores.
+
+### Non-Goals
+
+* This enhancement does not cover how OSUS running in the disconnected environment gets the signatures and if the 
+signatures are validated during the mirroring process.
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster admin will have to add the verification URI to CVO.
+2. CVO will reach the newly added URI to verify the release images.
+
+### API Extensions
+
+* CVO will grow a new field, an array, which will be populated with the image verification URI.
+* This will be an array as it'll allow the admin to use multiple locally configured stores. 
+* Existing default signature stores will be replaced with custom stores.
+* Existing stores can be used along with custom stores by adding them manually.
+* Admins will still be able to use the existing config maps for signature verification.
+* CVO will check the release signatures in the local ConfigMap first and will reach in parallel to other signature stores.
+
+This enhancement proposes to modify the [api/config/v1/types_cluster_version.go](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go) as following,
+
+A new field `SignatureStores` will be added to the [`ClusterVersionSpec`](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L40-L96) structure:
+
+```go
+type ClusterVersionSpec struct {
+	...
+	// SignatureStores contains the upstream URIs to verify release signatures.
+	// By default, CVO will use existing signature stores if this property is empty.
+	// The CVO will check the release signatures in the local ConfigMaps first. It will search for a valid signature
+	// in these stores in parallel only when local ConfigMaps did not include a valid signature.
+	// Validation will fail if none of the signature stores reply with valid signature before timeout.
+	// Setting SignatureStores will replace the default signature stores with custom signature stores.
+	// Default stores can be used with custom signature stores by adding them manually.
+	//
+	// Items in this list should be a valid absolute http/https URI of an upstream signature store as per rfc1738.
+	// +kubebuilder:validation:Format=XValidation:rule="self.all(x, isURL(x))",message="signatureStores must contain only valid absolute URLs per the Go net/url standard"
```

Comment:
[verify runs][1] on openshift/api#1601 are grumpy about something in this line.

```
error running generator schemapatch on group config.openshift.io: 
	could not run schemapatch generator for group/version config.openshift.io/v1: 
		could not execute schemapatch for manifest /go/src/github.com/openshift/api/config/v1/0000_00_cluster-version-operator_01_clusteroperator.crd.yaml: 
			/go/src/github.com/openshift/api/config/v1/types_cluster_version.go:100:2: extra arguments provided: ":rule=\"self.all(x, isURL(x))\",message=\"signatureStores must contain only valid absolute URLs per the Go net/url standard\"" (at <input>:1:12)
```

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_api/1678/pull-ci-openshift-api-master-verify/1727037884873052160#1:build-log.txt%3A14124-14128

Reply:
Possibly we need to drop `Format=`?  I'm just pattern-matching off:

```console
openshift/api $ git grep 'XValidation.*self.all' | head -n1
config/v1/types_infrastructure.go:      // +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="resourceTags are immutable and may only be configured during installation"
```

Reply:
```suggestion
	// +kubebuilder:validation:XValidation:rule="self.all(x, isURL(x))",message="signatureStores must contain only valid absolute URLs per the Go net/url standard"
```

Reply:
fixed